WIP: Disable autostart for Loweheiser generator

Georacer opened the pull request:
# Summary

This PR removes support for auto-starting the Loweheiser generator.
Now the user has to assign a starter switch channel and start the engine during IDLE state.

# Details

The reason is that, according to the manufacturer and a commercial user, it's not 100% reliable to use the starter (which is optional, as an alternative to a ripcord) so it's preferable for the user to trigger it intentionally.

Note: This PR has been re-written and re-coded from scratch.

This is a follow-up to https://github.com/ArduPilot/ardupilot/pull/29587.

# Known issues

After this PR, the engine won't auto-start any longer if the engine stalls mid-flight.
Also the user can't re-start it in-flight either, as starting is allowed only in IDLE state, and state changes are not allowed while armed.

peterbarker reviewed (changes requested):
I would have thought this could have been a rather smaller patch.  Not sure why the re-arrangements are required, and they've removed the above checks.

Where are the new tests to make sure this continues to function?

Georacer commented on libraries/AP_Generator/AP_Generator_Loweheiser.cpp:
```diff
@@ -443,19 +436,49 @@ void AP_Generator_Loweheiser::command_generator()
+            starter_requested_user = false;
+            break;
+        case RC_Channel::AuxSwitchPos::HIGH:
+            // Request running the starer on a rising edge.
```
Done.

alfmh commented:
Hi @Georacer ,

In our case we can build the code for Cube Orange, pixhawk4 and pixhawk5x, but we can't get telemetry or send commands to the generator. Also we can't configure all params than in the old version that was working.

Could share  the repository where you are fixing the code for testing generator?

Best regards,
Alfonso

peterbarker commented on libraries/AP_Generator/AP_Generator_Loweheiser.cpp:
```diff
@@ -384,13 +390,91 @@ void AP_Generator_Loweheiser::command_generator()
     bool user_controlled_starter = false;
+    if (rc_channel_starter_motor != nullptr) {user_controlled_starter = true;}
+    if (rc().has_valid_input() &&
+        (user_controlled_starter) &&
```
```suggestion
        user_controlled_starter &&
```

peterbarker commented on libraries/AP_Generator/AP_Generator_Loweheiser.cpp:
```diff
@@ -384,13 +390,91 @@ void AP_Generator_Loweheiser::command_generator()
+    // Don't allow auto-starting in STOP state.
+    if (commanded_runstate == RunState::STOP) { auto_controled_starter = false;}
+    // Don't allow auto-starting on IDLE state if a user-channel has been configured.
+    if ((commanded_runstate == RunState::IDLE) && user_controlled_starter ) { auto_controled_starter = false;}
```
formatting

Georacer committed:
AP_Generator: Loweheiser: Don't auto engage IDLE state

Even if we don't have temperature data, the generator won't autostart,
so there's no point in switching to IDLE.

Also corrected comment

Georacer committed:
AP_Generator: Loweheiser: Removed autostarter logic

According to the manufacturer automatically engagning the starter motor
is not a reliable process. It should be done manually instead.

Also caught unhandled nan.

Georacer committed:
autotest: Delete Loweheiser test with auto starter

The functionality has been removed from the driver.

tridge commented on libraries/AP_Generator/AP_Generator.cpp:
```diff
@@ -46,9 +46,9 @@ const AP_Param::GroupInfo AP_Generator::var_info[] = {
+    // @Group: LH
     // @Path: AP_Generator_Loweheiser.cpp
-    AP_SUBGROUPVARPTR(_driver_ptr, "", 3, AP_Generator, backend_var_info),
+    AP_SUBGROUPVARPTR(_driver_ptr, "LH", 3, AP_Generator, backend_var_info),
```
that makes for things like LHRUNTIME which isn't great as a name

Georacer commented on libraries/AP_Generator/AP_Generator_Loweheiser.cpp:
```diff
@@ -384,13 +390,91 @@ void AP_Generator_Loweheiser::command_generator()
+    if (commanded_runstate == RunState::STOP) { auto_controled_starter = false;}
+    // Don't allow auto-starting on IDLE state if a user-channel has been configured.
+    if ((commanded_runstate == RunState::IDLE) && user_controlled_starter ) { auto_controled_starter = false;}
+    if (auto_controled_starter) {
```
This text no longer exists.

Georacer commented on libraries/AP_Generator/AP_Generator_Loweheiser.cpp:
```diff
@@ -170,12 +175,6 @@ bool AP_Generator_Loweheiser::should_emergency_stop()
 // generator is at and other information
 void AP_Generator_Loweheiser::update_runstate()
 {
-    // If the vehicle crashes then we assume the pilot wants to stop
-    // the motor. This is done as a once-off when the crash is
-    // detected to allow the operator to rearm the vehicle, or we end
-    // up in a catch-22 situation where we force the stop state on the
-    // generator so they can't arm and can't start the generator
-    // because the vehicle is crashed.
```
I don't see how this comment was correct before, so I removed it.
This check runs all the time (not one-off) and I don't understand why we should be arming a crashed vehicle.

Georacer commented on libraries/AP_Generator/AP_Generator_Loweheiser.cpp:
```diff
@@ -215,16 +219,6 @@ void AP_Generator_Loweheiser::update_runstate()
         break;
     }
 
-    if (isnan(packet.efi_clt)) {
```
I've talked with Loweheiser. The confirm that when you switch from STOP to IDLE, you power on the EFI.
If the EFI is powered off (for whatever reason), the generator system will still send back the Loweheiser MAVLink message, but the cylinder temperature values will be NaN.

However, they don't consider it a very safe action to automatically go to IDLE, so they don't recommend it.

I'll keep this automatic state transition out of this driver.
The user will have to keep an eye out for the "Waiting for EFI" statustext to identify this situation and manually cycle the generator state.

Georacer commented on libraries/AP_Generator/AP_Generator_Loweheiser.cpp:
```diff
@@ -443,19 +436,49 @@ void AP_Generator_Loweheiser::command_generator()
-    // then consider running the starter motor.  Run motor for 5s.
-    if (!user_controlled_starter && commanded_runstate != RunState::STOP) {
+    // Check if the user wants to run the starter.
+    if (rc_channel_starter_motor != nullptr) {
```
In the current PR iteration I the rc valid check for the starter is present.

Georacer reviewed (commented):
I've talked to Loweheiser on this PR. They told me that:

When you switch from STOP to IDLE, you power on the EFI. When you go back to STOP the Loweheiser MAVLink msg will still get received, but a lot of values will be stuck.

When the EFI is powered off (for whatever reason, not just because the user wants to), the Loweheiser MAVLink msg will contain NaN in many fields, with the cylinder temperature being the most important.

When this happens while disarmed, they think it's risky to automatically go back to IDLE. Something has gone wrong and it needs user attention.

When this happens in the air (while on RUN), automatically going back to IDLE won't save the aircraft, since IDLE doesn't produce power. The pilot needs to land as soon as possible, something has gone wrong.

The same goes for mid-flight restarts, they are not an option currently when it comes to this specific hardware. Maybe in the future.

@alfmh please review what I wrote above, in case I misremembered our conversation.
